File: js/src/maplestory/api/dto/character/characterSetEffectSetDto.ts
import { CharacterSetEffectInfoDto } from './characterSetEffectInfoDto';
import { CharacterSetEffectOptionFullDto } from './characterSetEffectOptionFullDto';
import { CharacterSetEffectSetDtoBody } from '../../response/character/characterSetEffectDtoBody';

/**
 * 세트 효과 정보
 */
class CharacterSetEffectSetDto {
  /**
   * 세트 효과 명
   */
  setName: string;

  /**
   * 세트 개수 (럭키 아이템 포함)
   */
  totalSetCount: number;

  /**
   * 적용 중인 세트 효과 정보
   */
  setEffectInfo: CharacterSetEffectInfoDto[];

  /**
   * 모든 세트 효과 정보
   */
  setOptionFull: CharacterSetEffectOptionFullDto[];

  constructor(obj: CharacterSetEffectSetDtoBody) {
    const { set_name, total_set_count, set_effect_info, set_option_full } = obj;

    this.setName = set_name;
    this.totalSetCount = total_set_count;
    this.setEffectInfo = set_effect_info.map(
      (info) => new CharacterSetEffectInfoDto(info),
    );
    this.setOptionFull = set_option_full.map(
      (info) => new CharacterSetEffectOptionFullDto(info),
    );
  }
}

export { CharacterSetEffectSetDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/kms/dto/character/CharacterSetEffectSetDTO.java
package dev.spiralmoon.maplestory.api.kms.dto.character;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.util.List;

/**
 * 캐릭터 세트 효과 정보
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@ToString
public class CharacterSetEffectSetDTO implements dev.spiralmoon.maplestory.api.common.dto.character.CharacterSetEffectSetDTO<
        CharacterSetEffectInfoDTO,
        CharacterSetEffectOptionFullDTO> {

    /**
     * 세트 효과 명
     */
    @JsonProperty("set_name")
    private String setName;

    /**
     * 세트 개수 (럭키 아이템 포함)
     */
    @JsonProperty("total_set_count")
    private long totalSetCount;

    /**
     * 적용 중인 세트 효과 정보
     */
    @JsonProperty("set_effect_info")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterSetEffectInfoDTO> setEffectInfo;

    /**
     * 모든 세트 효과 정보
     */
    @JsonProperty("set_option_full")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterSetEffectOptionFullDTO> setOptionFull;
}
